subprojects/signing/src/main/java/org/gradle/plugins/signing/signatory/internal/pgp/InMemoryPgpSignatoryProvider.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.plugins.signing.signatory.internal.pgp;
+
+import groovy.lang.Closure;
+import org.bouncycastle.openpgp.PGPSecretKey;
+import org.bouncycastle.openpgp.PGPUtil;
+import org.bouncycastle.openpgp.jcajce.JcaPGPSecretKeyRing;
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.Project;
+import org.gradle.plugins.signing.SigningExtension;
+import org.gradle.plugins.signing.signatory.SignatoryProvider;
+import org.gradle.plugins.signing.signatory.pgp.PgpSignatory;
+import org.gradle.plugins.signing.signatory.pgp.PgpSignatoryFactory;
+import org.gradle.util.ConfigureUtil;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.codehaus.groovy.runtime.DefaultGroovyMethods.asType;
+
+/**
+ * A {@link SignatoryProvider} of {@link PgpSignatory} instances read from
+ * ascii-armored in-memory secret keys instead of a keyring.
+ */
+public class InMemoryPgpSignatoryProvider implements SignatoryProvider<PgpSignatory> {
+
+    private final PgpSignatoryFactory factory = new PgpSignatoryFactory();
+    private final Map<String, PgpSignatory> signatories = new LinkedHashMap<>();
+    private final String defaultSecretKey;
+    private final String defaultPassword;
+
+    public InMemoryPgpSignatoryProvider(String defaultSecretKey, String defaultPassword) {
+        this.defaultSecretKey = defaultSecretKey;
+        this.defaultPassword = defaultPassword;
+    }
+
+    @Override
+    public PgpSignatory getDefaultSignatory(Project project) {
+        if (defaultSecretKey != null && defaultPassword != null) {
```

Comment:
⭕️ What if I don't have a password (bad idea, but possible afaik)

Reply:
`PgpSignatoryFactory` throws a `NullPointerException` for a `null` password. I added a test that verifies that one can pass an empty string to use a key without a password in 87cda60.